db/db_impl/db_impl_merge.cc

```diff
@@ -0,0 +1,361 @@
+      auto current = db->versions_->current_next_file_number();
+      if (current <= max_log_number) {
+        db->versions_->FetchAddFileNumber(max_log_number - current + 1);
+      }
+    }
+  }
+
+  // Unblock writes.
+  write_blockers.clear();
+
+  TEST_SYNC_POINT("DBImpl::MergeDisjointInstances:AfterMergeMemtable:1");
+
+  // # Merge table files
+  assert(s.ok());
+  autovector<VersionEdit> cf_edits;
+  cf_edits.resize(num_cfs);
+  for (size_t cf_i = 0; cf_i < num_cfs; cf_i++) {
+    auto* this_cfd = this_cfds[cf_i];
+    auto& edit = cf_edits[cf_i];
+    edit.SetColumnFamily(this_cfd->GetID());
+    for (size_t db_i = 0; db_i < db_impls.size(); db_i++) {
+      auto* super_version = cf_db_super_versions[cf_i][db_i];
+      if (super_version == nullptr) {
+        continue;
+      }
+      VersionStorageInfo& vsi = *super_version->current->storage_info();
+      auto& cf_paths = super_version->cfd->ioptions()->cf_paths;
+      auto SourcePath = [&](size_t path_id) {
+        // Matching `TableFileName()`.
+        if (path_id >= cf_paths.size()) {
+          assert(false);
+          return cf_paths.back().path;
+        } else {
+          return cf_paths[path_id].path;
+        }
+      };
+      const auto& target_path = this_cfd->ioptions()->cf_paths.front().path;
+      const uint64_t target_path_id = 0;
+      for (int level = 0; level < vsi.num_levels(); ++level) {
+        for (const auto& f : vsi.LevelFiles(level)) {
+          assert(f != nullptr);
+          const uint64_t source_file_number = f->fd.GetNumber();
+          const uint64_t target_file_number = versions_->FetchAddFileNumber(1);
+          std::string src = MakeTableFileName(SourcePath(f->fd.GetPathId()),
+                                              source_file_number);
+          std::string target =
+              MakeTableFileName(target_path, target_file_number);
+          s = GetEnv()->LinkFile(src, target);
+          if (!s.ok()) {
+            return s;
+          }
+          edit.AddFile(
+              level, target_file_number, target_path_id, f->fd.GetFileSize(),
+              f->smallest, f->largest, f->fd.smallest_seqno,
+              f->fd.largest_seqno, f->marked_for_compaction, f->temperature,
+              f->oldest_blob_file_number, f->oldest_ancester_time,
+              f->file_creation_time, f->file_checksum,
+              f->file_checksum_func_name, f->min_timestamp, f->max_timestamp);
+        }
+      }
+    }
+  }
+
+  // # Apply version edits
+  assert(s.ok());
+  {
+    autovector<autovector<VersionEdit*>> edit_ptrs;
+    autovector<const MutableCFOptions*> cf_mopts;
+    for (size_t i = 0; i < num_cfs; i++) {
+      edit_ptrs.push_back({&cf_edits[i]});
+      cf_mopts.push_back(this_cfds[i]->GetLatestMutableCFOptions());
+    }
+
+    auto old_capacity = table_cache_->GetCapacity();
+    if (merge_options.max_preload_files >= 0) {
+      // Refer to `LoadTableHandlers` for calculation details.
+      table_cache_->SetCapacity(
```

Comment:
If we share table_cache_ across dbs in the future (to solve too many open fds), this can be a pitfall.